### ISSUE TITLE
Model Extension as functors between models

### ISSUE THREAD

**Issue by jpfairbanks** [jpfairbanks/SemanticModels.jl#180]
If we treat the set of all julia types as a category where the objects are types and the morphisms are functions, then models are subcategories of this category. We can talk about metamodeling as constructing functors between subcategories.

I thought that we could say that a program transformation is a functor between models. `F:M->M'` but that doesn't work because you can't do things like split a type when extending a model. If you go the other way things make sense. We can say that a model `M'` extends a model `M` if there exists a functor `F:M'->M`. For example start with a program 

```julia
double(a) = 2a
function main(a::Int,b::Int)
    c = a*b
    d = double(c)
end
```

and transform it to the program
```julia
double(a) = 2a
function main(a::Int,b::Int)
    c = a*b^-1
    d = double(c)
end
```

drawn as subcategories of the julia types these programs are shown in the attached diagram
[modelfunctor1.pdf](https://github.com/jpfairbanks/SemanticModels.jl/files/3060517/modelfunctor1.pdf)

There is a functor from the category on the right to the one on the left given by

Ob(F)

|T | F(T)|
|--|-----|
|Z|Z|
|Z x Z| Z x Z|
|R | Z|

Morphisms part

|f | F(f)|
|--|-----|
|pi1 | pi1|
|pi2 | pi2|
|ab^-1 | ab|
|2* | 2* |

**Comment by jpfairbanks** [comment]
If the functor `F:M'->M` is a [full and faithful](https://en.wikipedia.org/wiki/Full_and_faithful_functors) does that make `M',M` equivalent categories? Is this an [equivalence relation](https://en.wikipedia.org/wiki/Equivalence_relation), namely is is reflexive, symmetric, and transitive? 

Does the set of full functors induce a poset on the category of models? Perhaps `M <= M'` if there exists a full functor `F:M'->M`. We need to prove that these properties derive from the definition of full functor.

**Comment by jpfairbanks** [comment]
I think that model transformations that "just change the math" without changing "the structure" of the model are fully faithful functors.

For example, this program transformation just changes math, without changing the structure. It should be some special type of functor like fully faithful or something.
```julia
ex′ = quote
    G = 6.754
    function force(m1, m2, r)
        (G * (m1 + m2)) / double(r)
    end
    double(a) = begin
            2a
        end
    x = 3
    y = 1.6
    r = 2.1
    F = force(x, y, r)
end
--------------------------
ex2′ = quote
    G = 6.754
    function force(m1, m2, r; )
        (G * (m1 * m2)) / double(r)
    end
    function double(a; )
        a ^ 2
    end
    x = 3
    y = 1.6
    r = 2.1
    F = force(x, y, r)
end
==========================
function force(m1, m2, r; )
    (G * (m1 * m2)) / double(r)
end
--------------------------
function force(m1, m2, r)
    (G * (m1 + m2)) / double(r)
end
==========================
function double(a; )
    a ^ 2
end
--------------------------
function double(a)
    2a
end
==========================
eval(ex) = 7.3972380952380945
--------------------------
eval(ex2) = 7.3512925170068035
==========================
```

**Comment by infvie** [comment]
I'm not so sure if I agree with "_just change the math_" I don't think that's well defined. For example if I change a 3-mean to a 4-mean cluster is that just a change of math? Structural the program won't change that. Another example issue I see is regression going from 'SLR' to 'Quadratic Regression' has semantic meaning to the operation itself. I think here we first need to restrict this to *primative* which we need to more rigiously define

**Comment by jpfairbanks** [comment]
Yeah "just change the math" is definitely not well defined. I think we can take the fully functor definition, which is a well defined concept and use that to define "same structure". Also, what do you mean by *primitive* in this context?

**Comment by infvie** [comment]
To me, I would consider any operation that adds constants which can be categorized as `affine` transformations to the model would be considered _primitive_.

Perhaps we could try and generalize further to have `x' is said to have a primitive transformation iff x is a proper subset` with the idea being a larger transformation _would have to alter the set x_ although I feel less certain of this
